api_clients/thangs_events.py

```diff
@@ -75,8 +77,71 @@ def _send_amplitude_event(self, event_name, event_properties):
         except Exception as e:
             print(e)
 
+    def get_deviceId(self):
+        try:
+            if self.__cached_deviceId != None:
+                return self.__cached_deviceId
+            
+            user_config_path = bpy.utils.resource_path('USER')
+            device_id_json_file_location = os.path.join(os.path.dirname(user_config_path), 'thangs-blender-addon-id.json')
+
+            if os.path.exists(device_id_json_file_location):
+                deviceId_file = open(device_id_json_file_location)
+                data = json.load(deviceId_file)
+                self.__cached_deviceId = data["deviceId"]
+                return self.__cached_deviceId
+            else:
+                folder_above_user_config = os.path.basename(os.path.dirname(user_config_path))
+                
+                if folder_above_user_config.casefold() == 'blender':
+                    self.__cached_deviceId = str(uuid.uuid4())
+                    deviceId = {
+                        'deviceId': self.__cached_deviceId,
+                    }
+                    with open(device_id_json_file_location, 'w') as json_file:
+                        json.dump(deviceId, json_file)
+                    return self.__cached_deviceId
+                else:
+                    error = "Could not find blender folder in this path:" + user_config_path
+                    send_amp_event_on_initialization_error(error)
+                    self.__cached_deviceId = socket.gethostname().split(".")[0]
+                    return self.__cached_deviceId
+        except Exception as e:
+            print(e)
+            send_amp_event_on_initialization_error(e)
+            self.__cached_deviceId = socket.gethostname().split(".")[0]
+            return self.__cached_deviceId
+
 __thangs_events__: ThangsEvents = None
 
+def _send_amp_event_on_initialization_error(error):
+    event_name = "Thangs Blender Addon - DeviceId Initialization Failed"
+    event = {
+        'event_type': event_name,
+        'device_id': socket.gethostname().split(".")[0],
+        'event_properties': {
+            'addon_version': str(get_config().version),
+            'device_os': str(platform.system()),
+            'device_ver': str(platform.release()),
+            'source': "blender",
+            'exception': str(error),
+        }
+    }
+
+    try:
+        response = requests.post(get_config().thangs_config['event_url'], json={'events': [event]})
+        log.info('Sent amplitude event: ' + event_name + 'Response: ' + str(response.status_code) + " " +
+                    response.headers['x-cloud-trace-context'])
+    except Exception as e:
+        print(e)
+
+def send_amp_event_on_initialization_error(error):
+    threading.Thread(
+        target=_send_amp_event_on_initialization_error,
+        args=(error,)
+    ).start()
+    return
+    
```

Comment:
These two send amp event functions are somewhat duplicates because the `ThangsEvents` object isn't actually in memory yet. When these function are called we are actually in the process of instantiating `ThangsEvents` and it's `deviceId` field.